Add protocol to allow different token strategies

chrisbetz opened the pull request:
Hi James,

sorry, for taking some time and bothering you again.

Here is a new version of the PR #16.


You need to protect yourself against CSRF-attacks if you store authentication information in cookies, as these will be sent along requests no matter whether their origin is legit or not. To cope with this, we do have different patterns, see https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet.

To implement a CSRF-strategy, where one does not need to share state between running server instances (not even via encrypted cookies), you need to resort to different state management strategies.

So far, only a random string was generated as csrf-token, stored to
session and compared to info from requests following
(https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet#Synchronizer_.28CSRF.29_Tokens).

Now, the user of ring-anti-forgery can decide on the strategy to handle
state.

This commit/PR

- adds both encrypted-token and signed-token, both relying on
  encryption to validate the csrf-token
- extends the API (in a backwards-compatible way) to
  specify a state management strategy
- changes the API (in a backwards compatible way) to get the CSRF-token
  (as it is costly and mostly unnecessary to generate new tokens for
  every request and also error prone to just keep old tokens from
  a session)


I do understand your desire to keep the project as slim as possible not to break existing clients, I'd suggest you review this PR first (keeping all in one place) and after that, we discuss the best strategies to split the project into the "core", aka a project providing only the existing session-based state management strategy and one or more other projects (sub-projects maybe, I'm a big fan of lein monolith ;) ) to provided different strategies.

That way, everybody should be happy, I hope.

BTW: I also patched ring-oauth to work without session, if you're interested into that, too. You might want to have a parallel look at https://github.com/gorillalabs/ring-oauth2/tree/feature_different-state-management-strategies.

Cheers,

Chris

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ring-clojure/ring-anti-forgery/17)
<!-- Reviewable:end -->

weavejester commented:
Thank you for working on this, however this seems very complex for what we want to do.

I'm confused why you've included public key encryption, and why you've included both JWE and JWS as a strategy. Why use Buddy at all when we can piggyback off the work done in Ring's cookie-store namespace?

chrisbetz commented:
Good morning!

Ah, I see. I'll try to clarify my requirements to make sure the motive for all this becomes clear.

First of all, the need for extending the behavior of ring-anti-forgery arose from my requirement to implement CSRF-mitigation for several services (running on different machines, so no shared state). Thus, I need to create a CSRF-token in one service (Service I, short for issuer) and validate it in another one (Service C, the validator). These services may come from different providers/maintainers.

To do so without leaking the secret between services (aka sharing with different people) I needed asymmetric signing of the CSRF-token. Only with asymmetric encryption, I'm able to sign the CSRF-Token in Service I with a private key and make the public key, well, public for other services to validate the token.

That's the reason to implement the signed-token-strategy.

Implementing the encrypted-token-strategy then was just a finger exercise, I do not need that one. It's "just" for simplifying things by relying on symmetric encryption and a simple key-phrase instead of public-/private-keypair. However, it still is "cookie-less" and relies on a single token embedded into the DOM (as a hidden field).


Using the cookie-store instead would also work and be a good idea to create a double-submit cookie strategy from that (https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet#Double_Submit_Cookie). But, as the OWASP-site states: if UI and functional services are split over several hosts, this might not be the way to go. The encrypted-token-strategy I added does not share this type of problem.

Hope making the requirement (different services for issuing and validation with no shared secret, generation of different parts of the site on different hosts) clarifies the motive for the signed-token-strategy.

I know it's a lot of hassle to put on you to understand and verify the PR. Thank you so much for all the effort you put in.

Cheers,

Chris

weavejester commented:
Thanks for the explanation! I think that might be a fairly unusual requirement, so I'd like to keep the protocol but push the buddy-based strategies out into another library. I don't think we need to burden the base library with all the additional dependencies when they're not required by most users.

Regarding the protocol, can you explain why the `delay-token-creation` is a necessary method? Why not assume that the token can always be delayed if the correct option is set?

I was considering a strategy protocol more like:

```clojure
(defprotocol Strategy
  (create-token [strategy])
  (read-token   [strategy request])
  (valid-token? [strategy token request])
  (write-token  [strategy token response]))
```

chrisbetz commented:
Hi,

I'm totally d'accord with your proposal: Will change the project layout to have a different project for the other state management strategies.

Will also simplify the protocol. Regarding `delay-token-creation`: That's just a simplification to make the option de facto unnecessary: While session-based state management should keep its default of non-delayed token creation, the encryption based strategies should have a delayed default. I envisioned this to be simpler to use for developers wanting to use ring-anti-forgery.

I need to think a little whether find-and-create and create can successfully be merged into a single function. Should work, but maybe I'll come back to you on that one ;)

... Next update will follow ... (got some other appointments today).

Cheers,

Chris

chrisbetz commented:
Hi,

I changed the project according to your suggestions, and moved all the encryption based stuff to a different repository (https://github.com/gorillalabs/ring-anti-forgery).

To check whether everything builds and runs fine, I set-up travis and clojars for different project coordinates (see https://travis-ci.org/gorillalabs/ring-anti-forgery and https://travis-ci.org/gorillalabs/ring-anti-forgery-strategies for Travis).

That way, you should be safe to merge this request. Please check the README for any changes you might want to apply.

If we're done, I could transfer ownership of https://github.com/gorillalabs/ring-anti-forgery-strategies to you, if you wish to co-develop ring-anti-forgery and the encryption-based strategies. Otherwise, I'll keep the strategies and manage them. Just as you like.

I really hope this change suits you plans.

Happy hacking :)

Chris

chrisbetz commented:
Hi,

regarding create-token and read-token: It just wasn't necessary to distinguish the two methods. So I just collapsed them into one, thinking "token" as a way to get the token (which might be an existing one or a new one, depending on the implementation and the context) was a good name. Could change that easily.

The DelayTokenCreation is just a marker for what's the best strategy to get a token: For the 'just generate a random string' version, things could (and should) stay the same. For the encryption based stuff, computation should be delayed. I think we agree on that and that's not your point.

I personally prefer making different behaviour of a library due to different configuration transparent. So, if I want to switch from using session-based tokens to encryption-based tokens, I should be aware that I'd better change from eager to lazy token creation - and to use `deref` or `@` when referring to the token in `*anti-forgery-token*`.

Yes, technically this is unnecessary, it's just a way to help others cope with change / the transition from one strategy to another. To make delaying explicit by allowing an option will give users the possibility to move from session based to encryption based strategy without touching any other code, if he decides to stick with eager token creation.

To simplify things for users who are just using one strategy, I'd like to provide a useful default for the delay-option. And that default should depend on the strategy used. Thus, I need to "mark" the strategy somehow.

I settled for the DelayTokenCreation protocol, because it reduces the use of purely technical stuff one needs to care ... like the method used before.

Regarding formatting: I just ran lein cljfmt fix, hope that suits you ;) Sorry I didn't before.

weavejester commented:
> regarding create-token and read-token: It just wasn't necessary to distinguish the two methods. So I just collapsed them into one

If we combine the methods, then we have duplicate code in every protocol implementation. If we separate the methods, we can write the following expression once:

```clojure
(or (read-token strategy request) (create-token strategy))
```

In general, we should prefer methods that are simple and avoid duplication.

> The DelayTokenCreation is just a marker for what's the best strategy to get a token: For the 'just generate a random string' version, things could (and should) stay the same.

I wonder if it wouldn't be better to remove the option from the middleware altogether, and instead rely on `force`.

So the default session strategy remains the same, and people can use the token directly, ensuring backward compatibility. However, going forward, other strategies may delay the token. Therefore future use should always `force` the bound token before use.

> Regarding formatting: I just ran lein cljfmt fix, hope that suits you ;) Sorry I didn't before.

That does fix some of the issues, but not all. For example, you still have lines that exceed 80 characters, multiple blank lines, and formatting changes that affect unrelated code. cljfmt doesn't solve all these issues yet.

However, we can tidy the commit up after it's all working to our mutual satisfaction :)

chrisbetz commented:
> The DelayTokenCreation is just a marker for what's the best strategy to get a token: For the 'just generate a random string' version, things could (and should) stay the same.
> 
> I wonder if it wouldn't be better to remove the option from the middleware altogether, and instead rely on force.
> 
> So the default session strategy remains the same, and people can use the token directly, ensuring backward compatibility. However, going forward, other strategies may delay the token. Therefore future use should always force the bound token before use.
> 
(facepalm) Now I get you. Yes, you’re right. Will change that :) Sorry, didn’t understand you before.

> regarding create-token and read-token: It just wasn't necessary to distinguish the two methods. So I just collapsed them into one
> 
> If we combine the methods, then we have duplicate code in every protocol implementation. If we separate the methods, we can write the following expression once:
> 
> (or (read-token strategy request) (create-token strategy))
> In general, we should prefer methods that are simple and avoid duplication.
> 
Actually, I don’t see the duplication from session-based token to signing/encryption-based token. They just do not share code. Otherwise, I’d implement a read-token method constantly returning nil on the signing/encryption-based implementations.

If you still prefer read-token and create-token, I’ll do it. However, I don’t get the point here.

> Regarding formatting: I just ran lein cljfmt fix, hope that suits you ;) Sorry I didn't before.
> 
> That does fix some of the issues, but not all. For example, you still have lines that exceed 80 characters, multiple blank lines, and formatting changes that affect unrelated code. cljfmt doesn't solve all these issues yet.
> 
> However, we can tidy the commit up after it's all working to our mutual satisfaction :)
> 
> 
Totally agreed.

weavejester commented:
> Actually, I don’t see the duplication from session-based token to signing/encryption-based token. They just do not share code.

Ah, it's because `read-token` is an option. I'd forgotten about that entirely.

Let me think about the best way to proceed. It might be worth deprecating that option.

weavejester commented:
I think we can deprecate the `:read-token` option. We keep it around and functional, but mark the option as deprecated in the docstring.

Instead, the strategy is responsible for reading, creating, validating and writing the token. If `:read-token` is set it can override the strategy, but we want to direct people to use the `:strategy` option instead.

chrisbetz commented:
Oh, attention: the read-token-option is reading the CSRF-token from the request (aka the one either added as header or as part of the form data). It is not the same as the `read-token` protocol idea  from `find-and-create-token`-function (which is reading it from the session).

So, if we move the `read-token` method to the strategy (which might be more coherent), one might also need to configure the read-token function. Thus, actually we're not bringing in more clarity or brevity by making this change.

It's more like a second protocol should be in place here, as this depends on the kind of application you're building - an SPA with requests from ClojureScript would put the CSRF-token in the header, while a classy form-based website would include it as hidden field aka as a form parameter. 

It's your call, all I ask you is to think about that change a second time.

BTW: I like the way to discuss things with you to come up with the best solution :)

weavejester commented:
Yes, you're right. I was getting mixed up.

So let's take another look at this. We certainly need a way of creating a token, since tokens that are signed are going to be different from tokens that are entirely random. So:

```clojure
(create-token strategy)
```

We need a way of verifying a token. Some strategies (like session and double-cookie) require access to the request. So:

```clojure
(valid-token? strategy request token)
```

Finally, some strategies will have to add the token to the response in some secondary way. For instance via a cookie or the session:

```clojure
(write-token strategy response token)
```

We leave `read-token` as a separate thing - as you correctly point out, it's independent of the strategy.

Have I missed out anything, or have I gotten it right this time?

chrisbetz commented:
> So let's take another look at this. We certainly need a way of creating a token, since tokens that are signed are going to be different from tokens that are entirely random. So:
> 
> (create-token strategy)
Yeees-and-no: In the session-strategy, you’re relying on a session-token, i.e. a token which stays the same as long as the user session is valid. So technically, its not creating a new token on every call to that method, but returning a session token if one exists and creating a new token otherwise.

That’s the reason I chose `token` as a name. But maybe it’s cleaner to stick with `find-or-create-token`.

However, due to necessity to retrieve the token from the session (in the session-based strategy), we need to add `request` as an argument. (We need this argument below, so let’s refer to this by session-token).

The request here is also necessary as the token may/should refer to information from the request like the logged in user’s ID to make sure a token is invalidated if one user logs out and another user logs in. See `get-subject-fn`in the `EncryptedTokenSMS`. Same is true for validation. So it has to be

```clojure
(create-token strategy request)
```
> We need a way of verifying a token. Some strategies (like session and double-cookie) require access to the request. So:
> 
> (valid-token? strategy request token)
Yes, I could move `(read-token request)` to the calling code and have token as a parameter here. No problem with that.
> Finally, some strategies will have to add the token to the response in some secondary way. For instance via a cookie or the session:
> 
> (write-token strategy response token)
As you do add a (new) token to the session only if it differs from the existing token, we need to have the prior token at hand (see session-token). So again: we need the request here:

```clojure
(write-token strategy response request token)
```
> We leave read-token as a separate thing - as you correctly point out, it's independent of the strategy.
> 
> Have I missed out anything, or have I gotten it right this time?
> 

If you find that acceptable, I’ll refactor to extract `read-token` from the strategies and have `token` as a parameter to `valid-token?` instead.

And just let me know if you prefer `create-token` or `find-or-create-token` over `token`.

Third thing: Maybe it’s nicer to re-sort the parameters to write-token so it’s (write-token strategy response request token)


So, to bring it together, we should have this protocol in the end:

```clojure
(find-or-create-token strategy request)
(valid-token? strategy request token) ;; where token is the one from the request
(write-token strategy request token response) ;; where token is a potentially delayed token to be written (new or from session)
```

Thank you :)

weavejester commented:
> Yeees-and-no: In the session-strategy, you’re relying on a session-token, i.e. a token which stays the same as long as the user session is valid.

I was considering regenerating the token each time, but on reflection this would cause problems if a person had multiple tabs loading at the same time, or multiple XHRs.

May I suggest `get-token` then as a generic name? I don't like `find-or-create-token` as a name, since when using an encrypted token, there is no finding going on. I also don't like `token` as a name, as functions without verbs tend to imply a pure transformation of data. So:

```clojure
(get-token strategy request)
```

> As you do add a (new) token to the session only if it differs from the existing token, we need to have the prior token at hand (see session-token).

You're right, and that makes me wonder about having a `find-token` and `create-token` again. For example:

```clojure
(defprotocol Strategy
  (find-token [strategy request])
  (create-token [strategy])
  (valid-token? [strategy token1 token2])
  (write-token [strategy response token]))
```

For encrypted tokens, `read-token` could just return `nil`. The advantage of this approach is that we only need to read the additional token once, instead of for each method, but my instinct is that this is too specific to one workflow.

The alternative is:

```clojure
(defprotocol Strategy
  (get-token [strategy request])
  (valid-token? [strategy request token])
  (write-token [strategy request response token]))
```

I'm leaning toward the latter. What do you think?

chrisbetz commented:
I'm totally for that latter variant, because the first one
1. is very much focussed on having an existing token in the first place and
2. lacks the ability to create and validate a token with information from the request.

Thus, I'll change the protocol to the latter version for you to have a look at and see how we proceed.

chrisbetz commented:
Ok, here you go. Protocol changed, formatting corrected. Awaiting you review / comments.

have a nice evening :)

weavejester commented:
Thanks! I'll do a thorough review.

weavejester commented on README.md:
```diff
@@ -3,7 +3,10 @@
 
 Ring middleware that prevents [CSRF][1] attacks by via a
-randomly-generated anti-forgery token.
+randomly-generated anti-forgery [synchronizer token](https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet#Synchronizer_.28CSRF.29_Tokens)
```
Can you change this to a form like: `[synchronizer token][2]`

weavejester commented on README.md:
```diff
@@ -3,7 +3,10 @@
+randomly-generated anti-forgery [synchronizer token](https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet#Synchronizer_.28CSRF.29_Tokens)
+or by implementing an [encrypted token](https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet#Encrypted_Token_Pattern).
+
+Make sure to always use tls (https), here especially use it to prevent replay attacks!
```
Wrap at 70 characters for the README.

weavejester commented on src/ring/middleware/anti_forgery.clj:
```diff
@@ -35,21 +17,15 @@
       (= method :options)))
 
-(defn- valid-request? [request read-token]
+(defn- valid-request? [state-management-strategy request read-token]
```
Change `state-management-strategy` to `strategy`. There's only one type of strategy, so we don'tneed to be overly verbose.

weavejester commented on src/ring/middleware/anti_forgery.clj:
```diff
@@ -87,23 +63,27 @@
   :error-handler  - a handler function to call if the anti-forgery token is
                     incorrect or missing.
 
+  :state-management-strategy - A state management strategy,
```
Change this key to `:strategy`. It's more concise without losing the meaning.

weavejester commented on src/ring/middleware/anti_forgery.clj:
```diff
@@ -87,23 +63,27 @@
    {:pre [(not (and (:error-response options) (:error-handler options)))]}
-   (let [read-token    (:read-token options default-request-token)
-         error-handler (make-error-handler options)]
+   (let [state-management-strategy (get options :state-management-strategy (session/->SessionSMS))
```
Align forms in `let`, keep lines to 80 characters or fewer.

weavejester commented on src/ring/middleware/anti_forgery/strategy.clj:
```diff
@@ -0,0 +1,24 @@
+(ns ring.middleware.anti-forgery.strategy)
+
+(defprotocol StateManagementStrategy
```
Change the protocol name to `Strategy`.

weavejester commented on src/ring/middleware/anti_forgery/strategy.clj:
```diff
@@ -0,0 +1,24 @@
+  of token creation and validation."
+
+  (get-token [strategy request]
+    "Returns a token to be used. Users of ring.middleware.anti-forgery should use the appropriate utility functions
```
Wrap at 80 characters.

weavejester commented on src/ring/middleware/anti_forgery/strategy/session.clj:
```diff
@@ -0,0 +1,36 @@
+(ns ring.middleware.anti-forgery.strategy.session
+  "Implements a synchronizer token pattern, see
+https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet#Synchronizer_.28CSRF.29_Tokens"
```
Let's use a URL shortener here: goo.gl/WRm7Kp

weavejester commented on src/ring/middleware/anti_forgery/strategy/session.clj:
```diff
@@ -0,0 +1,36 @@
+  strategy/StateManagementStrategy
+
+  (get-token [this request]
+    (or (session-token request)
```
Let's inline `session-token` and `create-token`. I don't think we make things clearer by movng them out into private functions.

chrisbetz replied:
session-token is used three times. I'd prefer to keep it.

weavejester replied:
Ah, you're right. I didn't notice the other uses.

weavejester commented on src/ring/middleware/anti_forgery/strategy/session.clj:
```diff
@@ -0,0 +1,36 @@
+
+  (valid-token? [_ request token]
+    (let [stored-token (session-token request)]
+      (and stored-token
```
This can be one line.

weavejester commented on src/ring/util/anti_forgery.clj:
```diff
@@ -1,11 +1,17 @@
+(defn anti-forgery-token []
+  (let [potentially-delayed-token anti-forgery/*anti-forgery-token*
+        token (force potentially-delayed-token)]
+    token))
```
No need to make it so verbose:

```clojure
(defn anti-forgery-token []
  (force anti-forgery/*anti-forgery-token*))
```

weavejester commented on test/ring/middleware/test/anti_forgery.clj:
```diff
@@ -1,8 +1,22 @@
-        ring.mock.request))
+  (:require  [clojure.test :refer :all]
+             [ring.middleware.anti-forgery :as af :refer :all]
+            [ring.middleware.anti-forgery.strategy :as strategy]
```
Form alignment is off.

weavejester commented on test/ring/middleware/test/anti_forgery.clj:
```diff
@@ -1,8 +1,22 @@
+;; Generic helpers
+;;
+
+(defn- status=* [handler status req]
```
This function is unused and can be removed.

weavejester commented on test/ring/util/test/anti_forgery.clj:
```diff
@@ -1,7 +1,7 @@
-  (:use clojure.test
-        ring.util.anti-forgery
-        ring.middleware.anti-forgery))
+  (:require [clojure.test :refer :all]
```
Save this for another PR. It's not related to the changes.

weavejester replied:
This still applies. We can take this namespace entirely out of the commit.

weavejester commented on src/ring/middleware/anti_forgery/strategy/session.clj:
```diff
@@ -0,0 +1,36 @@
+(defn- create-token [request]
+  (random/base64 60))
+
+(deftype SessionSMS []
```
Change the name to `SessionStrategy` and add in a public function `session-strategy`.

weavejester commented on project.clj:
```diff
@@ -13,4 +13,4 @@
    :1.7 {:dependencies [[org.clojure/clojure "1.7.0"]]}
    :1.8 {:dependencies [[org.clojure/clojure "1.8.0"]]}
-   :1.9 {:dependencies [[org.clojure/clojure "1.9.0-alpha11"]]}})
+   :1.9 {:dependencies [[org.clojure/clojure "1.9.0"]]}})
```
Remove change - it's not directly relevant to the PR so we can handle this in a separate commit.

weavejester commented on src/ring/middleware/anti_forgery/strategy/session.clj:
```diff
@@ -0,0 +1,36 @@
+(ns ring.middleware.anti-forgery.strategy.session
```
Move to `ring.middleware.anti-forgery.session` for consistency with `ring.middleware.session`.

weavejester commented on src/ring/middleware/anti_forgery/strategy/session.clj:
```diff
@@ -0,0 +1,36 @@
+  (get-in request [:session :ring.middleware.anti-forgery/anti-forgery-token]))
+
+(defn- add-session-token [session-sms response request token]
+  (if response
```
This check can be done outside the protocol.

weavejester commented on src/ring/middleware/anti_forgery/strategy/session.clj:
```diff
@@ -0,0 +1,36 @@
+(defn- session-token [request]
+  (get-in request [:session :ring.middleware.anti-forgery/anti-forgery-token]))
+
+(defn- add-session-token [session-sms response request token]
```
The body of this function can be moved into the `write-token` method.

weavejester commented on src/ring/middleware/anti_forgery.clj:
```diff
@@ -87,23 +63,27 @@
+        (if (valid-request? state-management-strategy request read-token)
+          (let [token (strategy/get-token state-management-strategy request)]
+            (binding [*anti-forgery-token* token]
+              (handler request #(respond (strategy/write-token state-management-strategy request % token)) raise)))
```
Keep at 80 characters or fewer.

weavejester commented on src/ring/middleware/anti_forgery/strategy.clj:
```diff
@@ -0,0 +1,24 @@
+
+  StateManagmentStrategy is the protocol to abstract the process
+  of token creation and validation."
+
```
Remove extraneous blank line.

weavejester commented on test/ring/middleware/test/anti_forgery.clj:
```diff
@@ -128,7 +142,7 @@
         AssertionError
         (wrap-anti-forgery (constantly {:status 200})
-                           {:error-handler (fn [request] {:status 500 :body "Handler"})
+                           {:error-handler  (fn [request] {:status 500 :body "Handler"})
```
This is just a formatting change, and while correct, we can drop it from this PR, since it's not a relevant change.

weavejester commented on README.md:
```diff
@@ -28,10 +45,12 @@ handler, inside of the standard `wrap-session` middleware in Ring:
-`*anti-forgery-token*` var.
+### Encrypted token to be used without session
+
+You can use other strategies to manage state (create, validitate and store the tokens), e.g. the encrypted token mode
```
Wrap at 70 characters, use 'footnote' style markdown links.

chrisbetz commented:
done. (Hopefully)

weavejester commented on src/ring/middleware/anti_forgery.clj:
```diff
@@ -1,30 +1,12 @@
+            [ring.middleware.anti-forgery.session :as session]))
 
-(def ^{:doc "Binding that stores an anti-forgery token that must be included
+(def ^{:doc     "Binding that stores an anti-forgery token that must be included
```
Some extra whitespace got added here.

weavejester commented on src/ring/middleware/anti_forgery.clj:
```diff
@@ -1,30 +1,12 @@
-        (-> response
-            (assoc :session (:session response (:session request)))
-            (assoc-in [:session ::anti-forgery-token] token))))))
+*anti-forgery-token*)
```
Some indentation got removed here - can you add it back in?

weavejester commented on src/ring/middleware/anti_forgery.clj:
```diff
@@ -87,23 +63,33 @@
+            (binding [*anti-forgery-token* token]
+              (handler request
+                       #(respond
+                          (when %
```
Indentation is one too far to the right. It should be:

```clojure
#(respond
  (when %
    (strategy/write-token strategy request % token)))
```

weavejester commented on src/ring/middleware/anti_forgery/session.clj:
```diff
@@ -0,0 +1,33 @@
+(ns ring.middleware.anti-forgery.session
+  "Implements a synchronizer token pattern, see
+   https://goo.gl/WRm7Kp"
```
Can be put on the previous line.

weavejester commented on src/ring/middleware/anti_forgery/session.clj:
```diff
@@ -0,0 +1,33 @@
+
+(deftype SessionStrategy []
+  strategy/Strategy
+
```
Can you cut out this extra newline?

weavejester commented on src/ring/middleware/anti_forgery/session.clj:
```diff
@@ -0,0 +1,33 @@
+              token))))))
+
+(defn session-strategy []
+  (->SessionStrategy))
```
Can you make sure the file ends in a newline?

weavejester commented on src/ring/middleware/anti_forgery.clj:
```diff
@@ -87,23 +63,33 @@
    {:pre [(not (and (:error-response options) (:error-handler options)))]}
-   (let [read-token    (:read-token options default-request-token)
-         error-handler (make-error-handler options)]
+   (let [strategy (:strategy options (session/session-strategy))
```
Can you align the `let` clauses so it's was the same as before?

weavejester commented on src/ring/middleware/anti_forgery.clj:
```diff
@@ -87,23 +63,33 @@
-         error-handler (make-error-handler options)]
+   (let [strategy (:strategy options (session/session-strategy))
+         read-token (:read-token options default-request-token)
+         error-handler-fn (make-error-handler options)]
```
Can you change this symbol back to `error-handler` as it was before? The `-fn` is redundant.

chrisbetz replied:
Ok, I just disliked shadowing clojure.core/error-handler. But that's another PR.

weavejester replied:
I think shadowing is fine so long as the meaning remains clear.

In this case, `clojure.core/error-handler` is such an obscure function that I doubt anyone would assume that `error-handler` referred to the var rather than a local symbol, especially in the context of Ring, where "handler" has a precise meaning.

chrisbetz replied:
Oh, that's fine. I'm already taking it back. Just wanted to explain why I switched in the first place.

weavejester commented:
Thanks! I just have a few more formatting fixes, but it's almost done.

Once those last few are cleared up, can you squash your commits down and change the commit message to something like:

```
Add :strategy option and protocol to middleware

This allows other CSRF-protection strategies to be employed, such as
encrypted tokens, or the double submit cookie pattern.

See: https://goo.gl/2u4rgB
```

chrisbetz commented:
Ok, hope you find that good :)

Have a nice evening.

chrisbetz commented:
Thanks :)

Would you mind updating the dependencies in ring-defaults or do you want me to create a PR for that? Would make my life a lot easier.
